More feature to betterpairing added and some refactoring performed

nqyy opened the pull request:
1. Added some new methods in classes from betterpairing: G1, G2.
2. Added a new class to betterpairing: GT.
3. On the rust side, new class PyFq6 is added to lib.rs
4. New rust method vec_sum is added.
5. Some refactoring is performed on betterpairing.py

nqyy commented on honeybadgermpc/betterpairing.py:
```diff
@@ -361,27 +496,25 @@ def __add__(self, other):
         elif type(other) is int:
             if other < 0:
-                other *= 1
+                other *= -1
```
I think this is a bug.

amiller replied:
~~`other *= -1` can be different than `other = other * -1`. 
The difference is `*=` applied to an object actually runs the `__imul__` operator, which for the `ZR` `G1` etc objects we have here, performs modifications in place. I suspect this is the cause of problems~~
That can't be it, as I can't find any examples where this is applied to anything except an int

nqyy commented on honeybadgermpc/betterpairing.py:
```diff
@@ -400,27 +533,25 @@ def __sub__(self, other):
         elif type(other) is int:
             if other < 0:
-                other *= 1
+                other *= -1
```
Same as above.

nqyy commented on honeybadgermpc/betterpairing.py:
```diff
@@ -134,32 +151,33 @@ def __setstate__(self, d):
+        self.pyg1.py_pairing_with(other.pyg2, fq12)
+        return GT(fq12)
+
     @staticmethod
```
I think we need ``@staticmethod``. Don't know why it was taken out.

nqyy commented on honeybadgermpc/betterpairing.py:
```diff
@@ -312,18 +335,127 @@ def rand(seed=None):
+
+    # TODO: __truediv__ needs to be implemented
+
+    # TODO: __pow__ needs to be implemented
```
``__truediv__`` and ``__pow__`` are not implemented. They are called below.

amiller replied:
I think we should narrow this interface abstraction a bit. G1 G2 and GT are both groups, but G1 and G2 are additive notation (you can add two points together, but not multiply them together).

   Now the slightly confusing thing is that GT is actually defined as the finite field Fq12.... and in general finite fields support multiplication as well as addition. However, the GT class should probably not expose `__add__` at all.

nqyy reviewed (commented):
Some comments by me.

amiller commented:
You have a few easy flake8 things to fix before you get the all greens :)

amiller commented on honeybadgermpc/betterpairing.py:
```diff
@@ -134,32 +151,33 @@ def __setstate__(self, d):
-        negone = PyFr(str(1))
-        negone.negate()
-        self.pyg1.mul_assign(negone)
+        self.pyg1.mul_assign(PyFr(str(1)).negate())
```
Maybe premature optimization - but this looks like an easy candidate for caching. 

Separately - It always irks me to see serialization/deserialization to decimal numbers as the way of communicating with an FFI -- we're encountering this in bindings to the C++ NTL library too so I don't have a solution in mind.

nqyy replied:
I see.

amiller commented on honeybadgermpc/betterpairing.py:
```diff
@@ -214,7 +235,8 @@ def __truediv__(self, other):
         else:
-            raise TypeError
+            raise TypeError(
+                'Invalid division param. Expected G1. Got ' + str(type(other)))
```
thanks, much more helpful

amiller commented on pairing/src/lib.rs:
```diff
@@ -540,10 +614,17 @@ impl PyFq12 {
 }
 
+#[pyfunction]
+fn vec_sum(a: &PyList) -> PyResult<String>{
```
nice, useful sum tool to expose

amiller commented on pairing/src/lib.rs:
```diff
@@ -474,6 +483,24 @@ impl PyFq2 {
 }
 
+#[pyclass]
+struct PyFq6 {
```
what do we need Fq6 exposed in python for?

amiller reviewed (changes requested):
Changes requested:
1. Clarify or justify why the exposed methods are appropriate, mainly I think addition in the GT group should be removed, and I don't see why Fq6 and Fq2 should be exposed
2. There are several new functions added but no tests, can we add those?

chitianhao commented:
Currently, there is no test about it at all. I'll add some to it later.

nqyy commented:
Flake8 issues are fixed. Some implementation reasons or details can be answered by @tyurek . Currently there is no test for betterpairing, so I am trying to write some tests for it.

amiller commented:
I'm OK to accept this and move on!

smkuls commented:
@chitianhao  Can you squash all the commits?

nqyy committed:
More feature to betterpairing added and some refactoring performed 

1. Added some new methods in classes from betterpairing: G1, G2.
2. Added a new class to betterpairing: GT.
3. On the rust side, new class PyFq6 is added to lib.rs
4. New rust method vec_sum is added.
5. Some refactoring is performed on betterpairing.py

nqyy commented:
Squashed and rebased. @smkuls

amiller commented on honeybadgermpc/betterpairing.py:
```diff
@@ -144,22 +165,25 @@ def duplicate(self):
-        out = PyG1()
-        out.zero()
-        return G1(out)
+        return G1(PyG1().zero())
```
This is the problem - `PyG1.zero()` zeroes out the element in place and returns `None`, rather than returning the `PyG1` interest.
